File: src/main/java/wraith/fabricaeexnihilo/modules/barrels/BarrelBlock.java
package wraith.fabricaeexnihilo.modules.barrels;

import net.fabricmc.fabric.api.transfer.v1.fluid.FluidConstants;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
import net.minecraft.block.BlockRenderType;
import net.minecraft.block.BlockState;
import net.minecraft.block.BlockWithEntity;
import net.minecraft.block.ShapeContext;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityTicker;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.ItemScatterer;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.biome.Biome;
import org.jetbrains.annotations.Nullable;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.ModEffects;
import wraith.fabricaeexnihilo.modules.barrels.modes.ItemMode;
import wraith.fabricaeexnihilo.modules.fluids.BloodFluid;
import wraith.fabricaeexnihilo.recipe.barrel.MilkingRecipe;

@SuppressWarnings({"UnstableApiUsage", "deprecation"})
public class BarrelBlock extends BlockWithEntity {
    private static final VoxelShape SHAPE = createCuboidShape(1.0, 0.0, 1.0, 15.0, 16.0, 15.0);

    private final boolean isFireproof;

    public BarrelBlock(Settings settings, boolean isFireproof) {
        super(settings);
        this.isFireproof = isFireproof;
    }

    @Nullable
    @Override
    public BlockEntity createBlockEntity(BlockPos pos, BlockState state) {
        return new BarrelBlockEntity(pos, state);
    }

    @Override
    public VoxelShape getOutlineShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return SHAPE;
    }

    @Override
    public BlockRenderType getRenderType(BlockState state) {
        return BlockRenderType.MODEL;
    }

    @Override
    public void precipitationTick(BlockState state, World world, BlockPos pos, Biome.Precipitation precipitation) {
        var blockEntity = world.getBlockEntity(pos);
        if (blockEntity instanceof BarrelBlockEntity barrel) {
            barrel.precipitationTick(precipitation);
        }
    }

    @Nullable
    @Override
    public <T extends BlockEntity> BlockEntityTicker<T> getTicker(World world, BlockState state, BlockEntityType<T> type) {
        return world.isClient ? null : checkType(type, BarrelBlockEntity.TYPE, BarrelBlockEntity::ticker);
    }

    public boolean isFireproof() {
        return isFireproof;
    }

    @Override
    public void onPlaced(World world, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack itemStack) {
        if (world.getBlockEntity(pos) instanceof BarrelBlockEntity barrelEntity) {
            EnchantmentHelper.get(itemStack).forEach((enchantment, level) -> barrelEntity.getEnchantmentContainer().setEnchantmentLevel(enchantment, level));
        }
    }

    @Override
    public void onStateReplaced(BlockState state, World world, BlockPos pos, BlockState newState, boolean moved) {
        if (!state.isOf(newState.getBlock()) && world.getBlockEntity(pos) instanceof BarrelBlockEntity barrel && barrel.getMode() instanceof ItemMode mode) {
            ItemScatterer.spawn(world, pos.getX(), pos.getY(), pos.getZ(), mode.getStack());
        }
        super.onStateReplaced(state, world, pos, newState, moved);
    }

    // Milking
    @Override
    public void onSteppedOn(World world, BlockPos pos, BlockState state, Entity entity) {
        var blockEntity = world.getBlockEntity(pos);
        if (blockEntity instanceof BarrelBlockEntity barrel && entity instanceof LivingEntity livingEntity) {
            if (FabricaeExNihilo.CONFIG.modules.barrels.enableBleeding) {
                var thorns = barrel.getEnchantmentContainer().getEnchantmentLevel(Enchantments.THORNS);
                if (thorns > 0
                        && barrel.fluidStorage.simulateInsert(FluidVariant.of(BloodFluid.STILL), 1, null) >= 1
                        && livingEntity.damage(world.getDamageSources().cactus(), thorns / 2F)) {
                    var amount = FluidConstants.BUCKET * thorns / livingEntity.getMaxHealth();
                    try (Transaction t = Transaction.openOuter()) {
                        barrel.fluidStorage.insert(FluidVariant.of(BloodFluid.STILL), (long) amount, t);
                        t.commit();
                    }
                }
            }
            if (!(livingEntity instanceof PlayerEntity) && !livingEntity.hasStatusEffect(ModEffects.MILKED) && FabricaeExNihilo.CONFIG.modules.barrels.enableMilking) {
                var recipe = MilkingRecipe.find(livingEntity.getType(), world);
                if (recipe.isPresent()) {
                    long inserted;
                    try (Transaction t = Transaction.openOuter()) {
                        inserted = barrel.fluidStorage.insert(recipe.get().getFluid(), recipe.get().getAmount(), t);
                        t.commit();
                    }
                    if (inserted > 0) {
                        livingEntity.addStatusEffect(new StatusEffectInstance(ModEffects.MILKED, recipe.get().getCooldown(), 1, false, false, false));
                    }
                }
            }
        }
        super.onSteppedOn(world, pos, state, entity);
    }

    @Override
    public ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hitResult) {
        if (world == null || pos == null) {
            return ActionResult.PASS;
        }
        var blockEntity = world.getBlockEntity(pos);
        return blockEntity instanceof BarrelBlockEntity barrelBlock
                ? barrelBlock.activate(player, hand)
                : ActionResult.PASS;
    }

}

File: src/main/kotlin/exnihilofabrico/modules/barrels/BarrelBlock.kt
package exnihilofabrico.modules.barrels

import alexiil.mc.lib.attributes.AttributeList
import alexiil.mc.lib.attributes.AttributeProvider
import alexiil.mc.lib.attributes.Simulation
import alexiil.mc.lib.attributes.fluid.volume.FluidVolume
import com.swordglowsblue.artifice.api.builder.data.recipe.ShapedRecipeBuilder
import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.registry.ExNihiloRegistries
import exnihilofabrico.modules.ModEffects
import exnihilofabrico.modules.barrels.modes.ItemMode
import exnihilofabrico.modules.base.BaseBlock
import exnihilofabrico.modules.base.addEnchantments
import exnihilofabrico.modules.fluids.BloodFluid
import exnihilofabrico.util.asEntity
import exnihilofabrico.util.asStack
import net.fabricmc.fabric.api.block.FabricBlockSettings
import net.minecraft.block.*
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.enchantment.Enchantments
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityContext
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.effect.StatusEffectInstance
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.SidedInventory
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.Identifier
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.registry.Registry
import net.minecraft.util.shape.VoxelShape
import net.minecraft.world.BlockView
import net.minecraft.world.IWorld
import net.minecraft.world.World

class BarrelBlock(val texture: Identifier,
                  val craftIngredient1: Identifier, val craftIngredient2: Identifier,
                  settings: FabricBlockSettings = FabricBlockSettings.of(Material.WOOD)):
        BaseBlock(settings), BlockEntityProvider, AttributeProvider, InventoryProvider {

    override fun getOutlineShape(state: BlockState?, view: BlockView?, pos: BlockPos?, entityContext: EntityContext?) = SHAPE
    override fun getRenderType(state: BlockState?) = BlockRenderType.MODEL

    override fun onUse(state: BlockState?, world: World?, pos: BlockPos?, player: PlayerEntity?, hand: Hand?, hitResult: BlockHitResult?): ActionResult {
        if(world?.isClient != false || pos == null)
            return ActionResult.CONSUME
        val blockEntity = world.getBlockEntity(pos)
        if(blockEntity is BarrelBlockEntity)
            return blockEntity.activate(state, player, hand, hitResult)
        return ActionResult.CONSUME
    }
    override fun addAllAttributes(world: World, pos: BlockPos, state: BlockState, attributes: AttributeList<*>) {
        val blockEntity = world.getBlockEntity(pos)
        if(blockEntity is BarrelBlockEntity) {
            attributes.offer(blockEntity.itemTransferable)
            attributes.offer(blockEntity.fluidTransferable)
        }
    }

    override fun getInventory(state: BlockState?, world: IWorld?, pos: BlockPos): SidedInventory? {
        return (world?.getBlockEntity(pos) as? BarrelBlockEntity)?.inventory
    }

    override fun onBreak(world: World, pos: BlockPos, state: BlockState, player: PlayerEntity?) {
        if(player?.isCreative == false)
            (world.getBlockEntity(pos) as? BarrelBlockEntity)?.let{barrel ->
                val stack = this.asStack()
                stack.addEnchantments(barrel.enchantments)
                world.spawnEntity(stack.asEntity(world, pos))
                (barrel.mode as? ItemMode)?.let { mode ->
                    world.spawnEntity(mode.stack.asEntity(world, pos))
                }
            }
        super.onBreak(world, pos, state, player)
    }

    /**
     * Milking
     */
    override fun onSteppedOn(world: World, pos: BlockPos, entity: Entity) {
        (entity as? LivingEntity)?.let{living ->
            if(ExNihiloFabrico.config.modules.barrels.enableBleeding) {
                (world.getBlockEntity(pos) as? BarrelBlockEntity)?.let { barrel ->
                    val thorns = barrel.enchantments.getEnchantmentLevel(Enchantments.THORNS)
                    if(thorns > 0 && living.damage(DamageSource.CACTUS, thorns.toFloat()/2)) {
                        val volume = FluidVolume.create(BloodFluid.still, (FluidVolume.BUCKET*thorns / living.maximumHealth).toInt())
                        barrel.fluidTransferable.attemptInsertion(volume, Simulation.ACTION)
                    }
                }
            }
            if(entity is PlayerEntity)
                return
            if(!living.hasStatusEffect(ModEffects.MILKED)) {
                val duration = ExNihiloRegistries.BARREL_MILKING.getResult(entity)?.let { (volume, cooldown) ->
                    (world.getBlockEntity(pos) as? BarrelBlockEntity)?.fluidTransferable?.attemptInsertion(volume, Simulation.ACTION)
                    cooldown } ?: 72000
                living.addStatusEffect(StatusEffectInstance(ModEffects.MILKED, duration,1, false, false, false))
            }
        }
    }

    /**
     * BlockEntity functions
     */
    override fun hasBlockEntity() = true
    override fun createBlockEntity(world: BlockView?) = BarrelBlockEntity(isStone = this.material == Material.STONE)

    override fun onPlaced(world: World?, pos: BlockPos?, state: BlockState?, placer: LivingEntity?, itemStack: ItemStack?) {
        val barrel = (world?.getBlockEntity(pos) as? BarrelBlockEntity) ?: return
        EnchantmentHelper.getEnchantments(itemStack).forEach { enchantment, level ->
            barrel.enchantments.setEnchantmentLevel(enchantment, level)
        }
    }

    fun generateRecipe(builder: ShapedRecipeBuilder) {
        builder.pattern("x x", "x x", "xyx")
            .ingredientItem('x', craftIngredient1)
            .ingredientItem('y', craftIngredient2)
            .result(Registry.ITEM.getId(asItem()), 1)
    }

    companion object {
        val SHAPE: VoxelShape = createCuboidShape(1.0, 0.0, 1.0, 15.0, 16.0, 15.0)
    }
}